research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+  selfplay_dir = os.path.join(FLAGS.base_dir, 'data/selfplay/')
+  holdout_dir = os.path.join(FLAGS.base_dir, 'data/holdout/')
+  training_chunk_dir = os.path.join(FLAGS.base_dir, 'data/training_chunks/')
+  sgf_dir = os.path.join(FLAGS.base_dir, 'sgf/')
+  evaluate_dir = os.path.join(FLAGS.base_dir, 'sgf/evaluate/')
+
+  print('Creating random initial weights...')
+  bootstrap(estimator_working_dir, models_dir, params)
+
+  # Start from the bootstrap model to play games.
+  print('Playing some games...')
+  model_name = '000000-bootstrap'
+  for rl_iter in range(params.max_iters_per_pipeline):
+    print('RL_iteration: %d' % rl_iter)
+
+    # Self-play to generate at least params.max_games_per_generation games
+    while True:
+      params.holdout_pct = 0.05
+      selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
+               params)
+      if FLAGS.holdout:
+        params.holdout_pct = 1
+        selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
+                 params)
+      games = tf.gfile.Glob(os.path.join(selfplay_dir, model_name, '*.zz'))
+      if len(games) >= params.max_games_per_generation:
+        break
+
+    print('Gathering game output...')
+    gather(selfplay_dir, training_chunk_dir, params)
+
+    print('Training on gathered game data...')
+    train(models_dir, estimator_working_dir, training_chunk_dir, params)
+
+    if FLAGS.holdout:
+      print('Validating on the holdout game data...')
+      validate(models_dir, holdout_dir, estimator_working_dir, params)
+
+    _, latest_model_name = model_utils.get_latest_model(models_dir)
+    if FLAGS.eval:  # Perform evaluation if needed
+      print('Evaluating the latest model...')
+      model_name = evaluate(
+          models_dir, model_name, latest_model_name, evaluate_dir, params)
+      print('Winner: %s!' % model_name)
+    else:
+      model_name = latest_model_name
+
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--base_dir',
+      type=str,
+      default='/tmp/minigo/',
+      metavar='BD',
+      help='Base directory for the MiniGo models and datasets.')
+  parser.add_argument(
+      '--go_size',
+      type=int,
+      default=9,
+      metavar='N',
+      choices=[9, 19],
+      help='Go board size. The default size is 9.')
+  parser.add_argument(
+      '--eval',
+      type=bool,
+      default=False,
+      metavar='EVAL',
+      help='A boolean to specify evaluation in the RL pipeline.')
+  parser.add_argument(
+      '--debug',
+      type=bool,
+      default=False,
+      metavar='DEBUG',
+      help='A boolean to indicate debug mode for testing purpose.')
+  parser.add_argument(
+      '--holdout',
+      type=bool,
+      default=False,
+      metavar='HOLDOUT',
+      help='A boolean to explicitly generate holdout data for validation.')
+
+  FLAGS, unparsed = parser.parse_known_args()
+  tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
```

Comment:
We will have to change this when going into official, though doesn't necessarily need to happen immediately.